components/homme/src/share/reduction_mod.F90

```diff
@@ -337,6 +342,92 @@ subroutine pmax_mt_r_1d(red,redp,len,hybrid)
   end subroutine pmax_mt_r_1d
 
 
+!max with location index
+!at the end result is in redp thread local array
+  subroutine ParallelMaxWithIndex(redp,hybrid)
+    use hybrid_mod, only : hybrid_t
+#ifdef _MPI
+    use parallel_mod, only: mpi_maxloc, mpi2real_t
+#endif
+
+    real (kind=real_kind), intent(inout) :: redp(2) ! thread private vector of partial sum
+    type (hybrid_t),       intent(in)    :: hybrid  ! parallel handle
+
+    ! Local variables
+    integer ierr, k
+
+    ! the first and fastest thread performs initializing copy
+    !$OMP SINGLE
+    red_max_index%buf(1:2) = redp(1:2)
+    red_max_index%ctr = hybrid%ithr
+    !$OMP END SINGLE
+
+    ! all other threads now do the max_op wrt the first thread's data
+    !$OMP CRITICAL (CRITMAXIND)
+    if (hybrid%ithr /= red_max_index%ctr) then
+       if (red_max_index%buf(1) < redp(1)) then
+          red_max_index%buf(1:2) = redp(1:2)
+       endif
+    end if
+    !$OMP END CRITICAL (CRITMAXIND)
+
+#ifdef _MPI
+    !$OMP BARRIER
+    if (hybrid%ithr==0) then
+       call MPI_Allreduce(red_max_index%buf(1),redp,1,MPI2real_t, &
+            MPI_MAXLOC,hybrid%par%comm,ierr)
+    end if
+#else
+    redp(1:2) = red_max_index%buf(1:2)
+#endif
+    !$OMP BARRIER
+
+  end subroutine ParallelMaxWithIndex
+
+
+!max with location index
+!result is in redp thread local array
+  subroutine ParallelMinWithIndex(redp,hybrid)
+    use hybrid_mod, only : hybrid_t
+#ifdef _MPI
+    use parallel_mod, only: mpi_minloc, mpi2real_t
+#endif
+
+    real (kind=real_kind), intent(inout) :: redp(2) ! thread private vector of partial sum
+    type (hybrid_t),       intent(in)    :: hybrid  ! parallel handle
+
+    ! Local variables
+    integer ierr, k
+
+    ! the first and fastest thread performs initializing copy
+    !$OMP SINGLE
+    red_min_index%buf(1:2) = redp(1:2)
+    red_min_index%ctr = hybrid%ithr
+    !$OMP END SINGLE
+
+    ! all other threads now do the max_op wrt the first thread's data
+    !$OMP CRITICAL (CRITMAXIND)
+    if (hybrid%ithr /= red_min_index%ctr) then
+       if (red_min_index%buf(1) > redp(1)) then
+          red_min_index%buf(1:2) = redp(1:2)
+       endif
+    end if
+    !$OMP END CRITICAL (CRITMAXIND)
+
+#ifdef _MPI
+    !$OMP BARRIER
+    if (hybrid%ithr==0) then
+       call MPI_Allreduce(red_min_index%buf(1),redp,1,MPI2real_t, &
+            MPI_MINLOC,hybrid%par%comm,ierr)
+!       red%buf(1:len)=redp(1:len)
```

Comment:
Need to do same changes for min version. Also, should remove this commented line.

Reply:
done.